### PR TITLE
[Catalog] Refactor secrets handling

### PR DESCRIPTION
**Pull request by snazy**
The handling of secrets is a bit awkward at the moment. We declare "IDs" on the objects, which are to be looked up via the `SecretsProvider`, which is currently just a `Map<String, String>`.

This change changes the behavior: The (default) behavior to lookup via another map is gone.

Secrets are now generally looked up in the following order:
1. If the value is defined on the "per bucket" config, it is used as is.
2. If a secret for the bucket can be resolved, it is used.
3. If the value is defined on the "default" config, it is used as is.
4. If a secret for the "default" config can be resolved, it is used.
5. The secret is assumed to not exist.

The above logic is implemented in the new `SecretsHelper` class. It calls the `SecretProvider` using a collection of keys to look up, which shall return a map with the existing, resolved secrets.

Follow-up, after #7043 is merged to `main`: refactor secrets handlings in "Nessie Core", where/if necessary.

### CODE REVIEW

**Comment by snazy** [comment]
The central piece of this PR is to remove the "artificial indirection" - how we later really resolve the secrets is a 2ndary concern IMO (aka: not used yet, so it can be changed ;) )

File: catalog/files/impl/src/main/java/org/projectnessie/catalog/files/secrets/SecretsHelper.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+   * <ol>
+   *   <li>The value of the attribute on the {@code specific} object,
+   *   <li>The resolved secret for the specific object (lookup via {@code baseName + '.' +
+   *       specificName + '.' + Specializeable.name()},
+   *   <li>The value of the attribute on the {@code base} object,
+   *   <li>The resolved secret for the base object (lookup via {@code baseName + '.' +
+   *       Specializeable.name()},
+   * </ol>
+   */
+  public static <R, B> B resolveSecrets(
```
Comment (adutra) [inline catalog/files/impl/src/main/java/org/projectnessie/catalog/files/secrets/SecretsHelper.java:49]: Hard to follow, even with the javadocs above. Looks slightly over-engineered, and I wonder if this will play nicely with Quarkus own secrets management: will this work with the Quarkus Vault extension? Or with the Quarkus Kubernetes Config extension?

Comment (adutra) [reply catalog/files/impl/src/main/java/org/projectnessie/catalog/files/secrets/SecretsHelper.java:49]: Also: the return value seems ignored everywhere.

File: catalog/files/impl/src/main/java/org/projectnessie/catalog/files/gcs/GcsOptions.java [hunk 2]
```diff
@@ -28,15 +37,52 @@ public interface GcsOptions<PER_BUCKET extends GcsBucketOptions> extends GcsBuck
...
+    GcsBucketOptions perBucket = buckets().get(name);
     if (perBucket == null) {
-      return this;
+      return resolveSecrets(name, null, secretsProvider);
+    }
+
+    return resolveSecrets(name, perBucket, secretsProvider);
+  }
+
+  Set<String> SECRET_NAMES =
```
Comment (adutra) [inline catalog/files/impl/src/main/java/org/projectnessie/catalog/files/gcs/GcsOptions.java:54]: Unused.

File: catalog/files/impl/src/main/java/org/projectnessie/catalog/files/secrets/SecretsProvider.java [hunk 3]
```diff
@@ -15,6 +15,23 @@
...
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
 public interface SecretsProvider {
-  String getSecret(String secret);
+  Map<String, String> resolveSecrets(Set<String> names);
+
+  static SecretsProvider mapSecretsProvider(Map<String, String> secretsMap) {
```
Comment (adutra) [inline catalog/files/impl/src/main/java/org/projectnessie/catalog/files/secrets/SecretsProvider.java:25]: Seems only used in tests.

**Review by adutra — APPROVED** [review]
Approving because I don't want to hold you any longer on this, but I find the overall design a bit convoluted.

**Comment by adutra** [comment]
Heads up: the file `docker/catalog-full/docker-compose.yml` needs to be changed accordingly.

File: catalog/files/impl/src/main/java/org/projectnessie/catalog/files/secrets/SecretsHelper.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+          secret.applicator().accept(builder, resolved);
+        }
+      }
+    }
+
+    return builder;
+  }
+
+  @NessieImmutable
+  public interface Specializeable<R, B> {
```
[the remark this answers is not in the document]
Comment (snazy) [reply catalog/files/impl/src/main/java/org/projectnessie/catalog/files/secrets/SecretsHelper.java:105]: I'm not sure (yet).

The only secrets-extension that can be "neatly" used in config files is the one for GCS. The other ones (AWS + Vault) need to be accessed programmatically, which is sad.

I was looking for a way to declare the "secrets options" but have a central place for the resolve-secrets-logic. Not yet sure how to integrate these three different secrets sources - especially wrt planned rotation and unplanned rotation (aka how to cache those).

[the remark this answers is not in the document]
Comment (snazy) [reply catalog/files/impl/src/main/java/org/projectnessie/catalog/files/secrets/SecretsHelper.java:105]: But that's likely for a different PR - I've started a WIP branch for that though.

**Review by adutra — APPROVED** [review]
